torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,120 @@
+import collections
+import warnings
+from functools import partial
+
+import torch
+from torch.testing._internal.common_cuda import (TEST_CUDA)
+import torch.testing._internal.common_methods_invocations as cmi
+from torch.testing._core import _dispatch_dtypes
+from torch.testing import (all_types_and_complex_and,
+                           all_types_and_complex,
+                           all_types_and_half,
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           floating_types_and_half,
+                           floating_types,
+                           integral_types,
+                           floating_types_and,
+                           floating_and_complex_types_and,
+                           integral_types_and,
+                           all_types_and,
+                           )
+
+COMPLETE_DTYPES_DISPATCH = (
+    all_types,
+    all_types_and_complex,
+    all_types_and_half,
+    floating_types,
+    floating_and_complex_types,
+    floating_types_and_half,
+    integral_types,
+    complex_types,
+)
+
+EXTENSIBLE_DTYPE_DISPATCH = (
+    all_types_and_complex_and,
+    floating_types_and,
+    floating_and_complex_types_and,
+    integral_types_and,
+    all_types_and,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+# Class to tag the dynamically generated types.
+
+
+class _dynamic_dispatch_dtypes(_dispatch_dtypes):
+    pass
+
+
+def get_supported_dtypes(op, sample_inputs_fn, device):
+    # Returns the supported dtypes for the given operator and device pair.
+    assert device in ['cpu', 'cuda']
+    if not TEST_CUDA and device == 'cuda':
+        warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA dtypes could be wrong")
+
+    supported_dtypes = set()
+    for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
```

Comment:
This is slightly incorrect because it's adding the dtype to supported dtypes after each sample input passes.

Maybe we should use this opportunity to use

https://github.com/pytorch/pytorch/blob/cab65ea3b94691af4e0ba530fa4aa070bc5c548e/test/test_ops.py#L54

directly. Can we call that test for each dtype and the appropriate device types, and use its success/failure as the signal we need?

Reply:
This nuance is mostly related to https://github.com/pytorch/pytorch/issues/58841
Since we make a set, we add the `dtype` once the argument supports it. Though if the behaviour by operator is appropriate. All samples should pass.

I'd ideally not like to depend on `test_supported_dtypes` (or any other actual test).
At this point test_supported_dtypes tests only one sample and thus is prone to not detecting https://github.com/pytorch/pytorch/issues/58841.

I think we can revisit this again.